File: backend/apps/members/utils/arabic.py
import re
import unicodedata

# Arabic diacritics (tashkeel/harakat, U+064B-U+065F and the extended
# Quranic annotation range U+0610-U+061A/U+06D6-U+06ED) plus tatweel/kashida
# (U+0640) — written as \u escapes rather than literal characters to keep
# this file unambiguous in any editor/encoding.
_TASHKEEL_RE = re.compile(
    "[ؐ-ًؚ-ٰٟۖ-ۭـ]"
)
_CHAR_MAP = str.maketrans(
    {
        "أ": "ا",  # أ -> ا
        "إ": "ا",  # إ -> ا
        "آ": "ا",  # آ -> ا
        "ٱ": "ا",  # ٱ -> ا
        "ة": "ه",  # ة -> ه
        "ى": "ي",  # ى -> ي
    }
)
_ARABIC_INDIC_DIGITS = str.maketrans("٠١٢٣٤٥٦٧٨٩", "0123456789")
_EXTENDED_ARABIC_INDIC_DIGITS = str.maketrans("۰۱۲۳۴۵۶۷۸۹", "0123456789")


def normalize_ar(text):
    """Normalize Arabic text for search matching.

    Strips diacritics/tatweel, unifies alef/teh-marbuta/alef-maqsura
    variants, collapses whitespace, casefolds. Store the result in
    Member.search_name and search against THAT column with a plain
    (engine-portable) LIKE — without this, a user typing "احمد" won't find
    a record stored as "أحمد" even though both refer to the same name, and
    they'll conclude search is broken.
    """
    if not text:
        return ""
    text = unicodedata.normalize("NFKC", text)
    text = _TASHKEEL_RE.sub("", text)
    text = text.translate(_CHAR_MAP)
    text = re.sub(r"\s+", " ", text).strip()
    return text.casefold()


def normalize_digits(text):
    """Convert Arabic-Indic (U+0660-U+0669) and Extended Arabic-Indic
    (U+06F0-U+06F9) digits to Latin (0-9). Phone keyboards set to an
    Arabic locale commonly produce these even though the UI displays
    Latin numerals throughout the app.
    """
    if not text:
        return text
    return text.translate(_ARABIC_INDIC_DIGITS).translate(_EXTENDED_ARABIC_INDIC_DIGITS)

File: frontend/src/lib/arabic.js
/** Client-side mirror of the backend's
 * `apps/members/utils/arabic.py::normalize_ar`.
 *
 * Kept deliberately in lockstep with that function: any client-side
 * filtering (the Combobox's option search) must match on the same
 * normalized form the server searches on, or the two disagree and a user
 * typing "احمد" gets "no results" locally for an option the server would
 * happily have matched as "أحمد". If you change one, change both.
 *
 * Every Arabic character here is written as a \u escape rather than a
 * literal. These are combining marks and RTL letters: as literals they
 * render invisibly or reorder in an editor, which makes the ranges
 * impossible to review and easy to corrupt on a careless edit.
 */

// Quranic annotation U+0610-U+061A, tashkeel/harakat U+064B-U+065F,
// superscript alef U+0670, Quranic annotation U+06D6-U+06ED,
// tatweel/kashida U+0640.
const TASHKEEL = /[ؐ-ًؚ-ٰٟۖ-ۭـ]/g;

// U+0623 alef-hamza-above, U+0625 alef-hamza-below, U+0622 alef-madda,
// U+0671 alef-wasla  -> U+0627 plain alef
// U+0629 teh-marbuta -> U+0647 heh
// U+0649 alef-maqsura -> U+064A yeh
const CHAR_MAP = {
  "أ": "ا",
  "إ": "ا",
  "آ": "ا",
  "ٱ": "ا",
  "ة": "ه",
  "ى": "ي",
};
const CHAR_MAP_RE = /[أإآٱةى]/g;

export function normalizeAr(text) {
  if (!text) return "";
  return String(text)
    .normalize("NFKC")
    .replace(TASHKEEL, "")
    .replace(CHAR_MAP_RE, (c) => CHAR_MAP[c])
    .replace(/\s+/g, " ")
    .trim()
    .toLowerCase();
}

// Arabic-Indic U+0660-U+0669, Extended Arabic-Indic U+06F0-U+06F9.
const ARABIC_INDIC = /[٠-٩]/g;
const EXTENDED_ARABIC_INDIC = /[۰-۹]/g;

/** Arabic-Indic and Extended Arabic-Indic digits to Latin (0-9). Mirrors
 * the backend's `normalize_digits`. The UI renders Latin numerals
 * everywhere, but an Arabic-locale phone keyboard still emits these.
 */
export function normalizeDigits(text) {
  if (!text) return text;
  return String(text)
    .replace(ARABIC_INDIC, (d) => String(d.charCodeAt(0) - 0x0660))
    .replace(EXTENDED_ARABIC_INDIC, (d) => String(d.charCodeAt(0) - 0x06f0));
}
